tests/ui/match/match-ref-mut-invariance.stderr

```diff
@@ -1,12 +1,12 @@
 error: lifetime may not live long enough
-  --> $DIR/match-ref-mut-invariance.rs:10:9
+  --> $DIR/match-ref-mut-invariance.rs:10:24
    |
 LL | impl<'b> S<'b> {
    |      -- lifetime `'b` defined here
 LL |     fn bar<'a>(&'a mut self) -> &'a mut &'a i32 {
    |            -- lifetime `'a` defined here
 LL |         match self.0 { ref mut x => x }
-   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ method was supposed to return data with lifetime `'b` but it is returning data with lifetime `'a`
+   |                        ^^^^^^^^^ assignment requires that `'a` must outlive `'b`
```

Comment:
oh, I missed that! I believe those bindings are "assignment"s because once they're lowered to MIR locals, they're assigned to by MIR assignment statements; the constraint generated by that is what's getting blamed. afaict currently there's not enough information in the MIR to efficiently make that distinction when assigning constraint categories. but we do have a span, so it shouldn't be too hard to call assignments "binding"s in diagnostics when syntactically they feel more like a "binding" than an "assignment".

that said, those are the wrong constraints to blame for these tests; these are both regressions, though it's just by chance they were correct in the first place. I assume because the error depends on `&mut` being invariant in its parameter, we're looking at things kind of backwards[^1] and we picked a constraint from the unhelpful end of the path. I think it should be possible to detect when constraints arising from contravariance are being followed and flip the path around if so.. it'll probably make for better diagnostics, but I'll admit I don't love relying so heavily on what direction we traverse the constraint graph in for this. I'm worried something more robust might require collecting more information in borrowck, but I'll look into it at least.

sorry for the huge number of very specific stderr changes, and thank you for taking the time to review this. I definitely agree that it's :/

[^1]: this is present in this test when blaming the correct constraint too: the label "method was supposed to return data with lifetime `'b` but it is returning data with lifetime `'a`" is peculiar when the return type in the signature has lifetime `'a` and the body has lifetime `'b`. I think that message was worded assuming return types are always covariant. it should be fixed too (maybe with a few other tweaks as well.. ideally it'd suggest replacing `'a` with `'b` rather than adding `'a: 'b`, like we do for `'static`), but probably that should go in a different PR.

Reply:
hmm, one approach would be to change this code here to consider more assignments to be boring: https://github.com/rust-lang/rust/blob/3378a5e084fa51c2e7f5c3f53b87e8ad89f1c517/compiler/rustc_borrowck/src/type_check/mod.rs#L1117-L1139 

E.g. by checking for `RefForGuard` and looking at `VarBindingForm::opt_match_place`. It's a bit fiddly sadly

Reply:
One likely quite impactful suggestion: make it boring if `opt_ty_info` is `None`

Reply:
that's where I was looking, yeah. though I was focusing too much on distinguishing between assignment syntaxes. `opt_match_place` does work for catching those by-ref bindings. and since we (probably?) don't need the span or place from that, checking the `VarBindingForm`'s `binding_mode` should be equivalent (I think). I don't think they should all be boring though, so there'd at need to be some other check (if not a different check entirely). e.g., for [field-projection-mutating-context.rs](https://github.com/rust-lang/rust/blob/master/tests/ui/mir/field-projection-mutating-context.rs):
```diff
1	error[E0308]: mismatched types
-	   --> $DIR/field-projection-mutating-context.rs:9:13
+	   --> $DIR/field-projection-mutating-context.rs:9:25
3	    |
4	 LL |     let Foo(ref mut y): Foo<fn(&'static str)> = x;
-	    |             ^^^^^^^^^ one type is more general than the other
+	    |                         ^^^^^^^^^^^^^^^^^^^^^ one type is more general than the other
6	    |
7	    = note: expected fn pointer `for<'a> fn(&'a _)`
8	               found fn pointer `fn(&_)`
```
the assignment is what relates the two types, so it makes more sense to blame than just one of the type annotations. there's also [fn-subtype.stderr](https://github.com/rust-lang/rust/blob/master/tests/ui/nll/relate_tys/fn-subtype.stderr) specifically checking for an assignment being blamed, though that assignment is by-value. in both cases the span associated with the assignment's constraint is unfortunate, but it is the assignment being blamed.

`RefForGuard` locals being boring to assign to makes sense, but it doesn't look like we blame those assignments in any tests anyway. I'd have to think more to come up with an example, if it's possible.

and `opt_ty_info` isn't *quite* what the docs claim. it looks like it's only `Some(_)` for function inputs (specifically for the inputs themselves, not projections from destructuring patterns). any other binding gets an `opt_ty_info` of `None` even if there's an annotation on it. we can get user annotations on normal local variables from a local decl's `user_ty` field, though. there's a slight asymmetry in how `UserTypeProjections` accounts nicely for projections and `opt_ty_info` doesn't, but also I can't imagine ever wanting to point to constraints from an argument being destructured, so that's probably fine (if unsettlingly ad-hoc). just checking for those also misses assignments from paths with explicitly-provided generic arguments, but maybe that's fine too; I think getting all user types and sorting out what's inferred from what's provided would be a much too involved process. in most of the tests this affects, it doesn't matter too much whether the assignment or something else gets blamed; they each provide half of the reason for a region error, so unless we surface more constraints' spans, the user has to hunt down the other half themself either way.